proposed/2021/MachineReadableOutput.md

```diff
@@ -0,0 +1,601 @@
+* Useful for CI/CD auditing(compliance, security ..)
+  * Produce Software Bill of Material (SBOM) (compliance, historic keeping)
+  * Enhancing Software Supply Chain Security
+    * Check any vulnerable packages
+    * Check any deprecated packages
+    * Check any outdated packages
+  * Check license compliance
+  * Resolve dependency issues (detect duplicate packages, newly introduced dependencies, dependency removals, etc.)
+  * Making the output machine-readable unlocks additional tooling and automation scenarios in CI/CD pipeline above scenarios.
+
+## Who are the customers
+
+Anyone (government/private enterprises, security experts, individual contributors) who wants to consume `dotnet list package` output for auditing tool or historic keeping, CI/CD orchestrating.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+#### `--format` option
+
+Ability to use new `--format` option for all `dotnet list package` commands to ensure formatted(JSON etc) output is emitted to the console.
+
+```dotnetcli
+dotnet list [<PROJECT>|<SOLUTION>] package [--config <SOURCE>]
+    [--deprecated]
+    [--framework <FRAMEWORK>] [--highest-minor] [--highest-patch]
+    [--include-prerelease] [--include-transitive] [--interactive]
+    [--outdated] [--source <SOURCE>] [-v|--verbosity <LEVEL>]
+    [--vulnerable]
+    [--format <FORMAT>]
+
+dotnet list package -h|--help
+```
+
+`<FORMAT>` - Allowed values as part of spec is `json`. (In the future `parseable`, `csv`, `yaml`, `xml` could be candidates.)
+
+#### `> dotnet list package`
+
+```dotnetcli
+Project 'MyProjectA' has the following package references
+   [netcoreapp3.1]:
+   Top-level Package                      Requested             Resolved
+   > Microsoft.Extensions.Primitives      [1.0.0, 5.0.0]        1.0.0
+   > NuGet.Commands                       4.8.0-preview3.5278   4.8.0-preview3.5278
+   > Text2Xml.Lib                         [1.1.2, 2.0.0)        1.1.2
+
+Project 'MyProjectB' has the following package references
+   [netcoreapp3.1]:
+   Top-level Package      Requested             Resolved
+   > NuGet.Commands       4.8.0-preview3.5278   4.8.0-preview3.5278
+   > Text2Xml.Lib         1.1.2                 1.1.2
+
+   [net5.0]:
+   Top-level Package      Requested             Resolved
+   > NuGet.Commands       4.8.0-preview3.5278   4.8.0-preview3.5278
+   > Text2Xml.Lib         1.1.2                 1.1.2
+```
+
+#### `> dotnet list package --format json`
+
+```json
+{
+  "version": 1,
+  "MyProjectA": [
+    {
+      "framework": "netcoreapp3.1",
+      "topLevelPackages": [
+        {
+          "id": "Microsoft.Extensions.Primitives",
+          "requestedVersion": "[1.0.0, 5.0.0]",
+          "resolvedVersion": "1.0.0"
```

Comment:
Should we include some sort of hash or package source used to restore the package? A package ID and version may have different content across different package sources. In other words, the package ID + version does not actually capture which package your project depends on.

Reply:
Keep in mind of dependency confusion attacks that .NET is susceptible to. We should consider making this information visible to help prevent these attacks

Reply:
While it's interesting from a "I want to know as much as possible" point of view, wouldn't customers that care about package hashes use lock files, and then they don't need write their own tooling to parse this command's json and handle it themselves?

In any case, I think it's scope creep for this spec (provide json output for existing features), so can be requested in a new issue, where customers can upvote and provide feedback about why it's important to them.